File: mycuration/src/main/java/com/phicdy/mycuration/presentation/presenter/TopActivityPresenter.java
package com.phicdy.mycuration.presentation.presenter;

import android.support.annotation.NonNull;
import android.view.MenuItem;

import com.phicdy.mycuration.R;
import com.phicdy.mycuration.data.db.DatabaseAdapter;
import com.phicdy.mycuration.presentation.view.TopActivityView;

public class TopActivityPresenter implements Presenter {

    private TopActivityView view;
    private DatabaseAdapter dbAdapter;
    private int launchTab = POSITION_CURATION_FRAGMENT;

    private static final int POSITION_CURATION_FRAGMENT = 0;
    private static final int POSITION_FEED_FRAGMENT = 1;
    private static final int POSITION_FILTER_FRAGMENT = 2;

    public TopActivityPresenter(int launchTab) {
        this.launchTab = launchTab;
    }

    public void setView(@NonNull TopActivityView view) {
        this.view = view;
    }

    @Override
    public void create() {
        view.initViewPager();
        view.initFab();
        view.setAlarmManager();
        view.changeTab(launchTab);
    }

    @Override
    public void resume() {
        new Thread(new Runnable() {
            @Override
            public void run() {
                dbAdapter.saveAllStatusToReadFromToRead();
            }
        }).start();

        view.closeSearchView();
    }

    @Override
    public void pause() {

    }

    public void setDataAdapter(DatabaseAdapter adapter) {
        this.dbAdapter = adapter;
    }

    public void fabClicked() {
        switch (view.currentTabPosition()) {
            case POSITION_CURATION_FRAGMENT:
                if (dbAdapter.getNumOfFeeds() == 0) {
                    view.goToFeedSearch();
                    return;
                }
                view.goToAddCuration();
                break;
            case POSITION_FEED_FRAGMENT:
                view.goToFeedSearch();
                break;
            case POSITION_FILTER_FRAGMENT:
                if (dbAdapter.getNumOfFeeds() == 0) {
                    view.goToFeedSearch();
                    break;
                }
                view.goToAddFilter();
                break;
            default:
        }
    }

    private void settingMenuClicked() {
        view.goToSetting();
    }

    public void optionItemClicked(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.setting_top_activity:
                settingMenuClicked();
                break;
            default:
        }
    }
}

File: mycuration/src/main/java/com/phicdy/mycuration/presentation/presenter/TopActivityPresenter.kt
package com.phicdy.mycuration.presentation.presenter

import android.view.KeyEvent
import android.view.MenuItem
import com.phicdy.mycuration.R
import com.phicdy.mycuration.data.repository.ArticleRepository
import com.phicdy.mycuration.data.repository.RssRepository
import com.phicdy.mycuration.presentation.view.TopActivityView
import kotlinx.coroutines.coroutineScope

class TopActivityPresenter(private val view: TopActivityView,
                           private val articleRepository: ArticleRepository,
                           private val rssRepository: RssRepository
) {

    fun create() {
        view.initViewPager()
        view.initFab()
        view.initToolbar()
        view.setAlarmManager()
    }

    suspend fun resume() = coroutineScope {
        articleRepository.saveAllStatusToReadFromToRead()
        view.closeSearchView()
    }

    fun fabClicked() {
        view.startFabAnimation()
    }

    suspend fun fabCurationClicked() = coroutineScope {
        view.closeAddFab()
        if (rssRepository.getNumOfRss() == 0) {
            view.goToFeedSearch()
            return@coroutineScope
        }
        view.goToAddCuration()
    }

    fun fabRssClicked() {
        view.closeAddFab()
        view.goToFeedSearch()
    }

    suspend fun fabFilterClicked() = coroutineScope {
        view.closeAddFab()
        if (rssRepository.getNumOfRss() == 0) {
            view.goToFeedSearch()
            return@coroutineScope
        }
        view.goToAddFilter()
    }

    fun addBackgroundClicked() {
        view.closeAddFab()
    }

    private fun settingMenuClicked() {
        view.goToSetting()
    }

    fun optionItemClicked(item: MenuItem) {
        when (item.itemId) {
            R.id.setting_top_activity -> settingMenuClicked()
        }
    }

    fun onKeyDown(keyCode: Int, isShowAddFabs: Boolean): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK && isShowAddFabs) {
            view.closeAddFab()
            return true
        }
        return false
    }

    fun queryTextSubmit(query: String?) {
        if (query == null) return
        view.goToArticleSearchResult(query)
    }
}
